File: practice_01/c++/BubbleSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <fstream>
#include <ctime>
#include <numeric>
#include <cmath>

// Implementación del algoritmo de burbuja en C++
void bubbleSort(int array[], int size) {
    for (int step = 0; step < size; ++step) {
        for (int i = 0; i < size - step - 1; ++i) {
            // Compara dos elementos adyacentes
            if (array[i] > array[i + 1]) {
                // Intercambia los elementos si no están en orden ascendente
                int temp = array[i];
                array[i] = array[i + 1];
                array[i + 1] = temp;
            }
        }
    }
}

int main() {
    // Tamaños de entrada para los experimentos
    std::vector<int> input_sizes = {100, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 40000, 50000};
    int num_repeats = 5; // Número de repeticiones para obtener el promedio

    std::ofstream outputFile;
    outputFile.open("bubble_sort_times.txt"); // Abrir archivo para escritura

    // Iterar a través de los diferentes tamaños de entrada
    for (int size : input_sizes) {
        std::vector<double> times; // Almacenar tiempos para repeticiones

        // Repetir el experimento varias veces
        for (int i = 0; i < num_repeats; ++i) {
            int data[size];
            srand(time(0));

            // Generar datos aleatorios para el tamaño de entrada actual
            for (int j = 0; j < size; ++j) {
                data[j] = rand() % size;
            }

            // Medir el tiempo antes y después de ejecutar el algoritmo
            auto start_time = std::chrono::high_resolution_clock::now();
            bubbleSort(data, size);
            auto end_time = std::chrono::high_resolution_clock::now();
            std::chrono::duration<double> elapsed_time = end_time - start_time;
            times.push_back(elapsed_time.count()); // Almacenar el tiempo en segundos
        }

        // Calcular el promedio y la desviación estándar de los tiempos
        double avg_time = std::accumulate(times.begin(), times.end(), 0.0) / num_repeats;
        double variance = 0.0;
        for (double time : times) {
            variance += std::pow(time - avg_time, 2);
        }
        double std_dev = std::sqrt(variance / num_repeats);
        
        // Escribir los resultados en el archivo
        outputFile << size << "\t" << avg_time << "\t" << std_dev << "\n";
    }

    outputFile.close(); // Cerrar archivo

    return 0;
}

File: practice_01/python/BubbleSort.py
import random
import time
import statistics

# Implementación del algoritmo de burbuja en Python
def bubbleSort(array):
    n = len(array)
    for i in range(n):
        # Realiza n-i-1 comparaciones en esta iteración, ya que los i elementos finales ya están en su lugar correcto
        for j in range(0, n - i - 1):
            # Compara los elementos adyacentes
            if array[j] > array[j + 1]:
                # Intercambia los elementos si no están en orden ascendente
                temp = array[j]
                array[j] = array[j + 1]
                array[j + 1] = temp

# Tamaños de entrada para los experimentos
input_sizes = [100, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 40000, 50000]
num_repeats = 5

with open("bubble_sort_times.txt", "w") as outputFile:
    # Iterar a través de los diferentes tamaños de entrada
    for size in input_sizes:
        times = [] # Almacenar tiempos para repeticiones

        print(f"Experimentando con tamaño de entrada: {size}")

        # Repetir el experimento varias veces
        for repeat in range(num_repeats):
            data = [random.randint(1, size) for _ in range(size)]

            print(f"Repetición {repeat + 1}/{num_repeats} - Datos generados: {data}")

            # Medir el tiempo antes y después de ejecutar el algoritmo
            start_time = time.time()
            bubbleSort(data)
            end_time = time.time()
            elapsed_time = end_time - start_time
            times.append(elapsed_time) # Almacenar el tiempo en segundos

            print(f"Repetición {repeat + 1}/{num_repeats} - Tiempo transcurrido: {elapsed_time:.6f} segundos")

        # Calcular el promedio y la desviación estándar de los tiempos
        if len(times) >= 2:
            avg_time = statistics.mean(times)
            std_dev = statistics.stdev(times)
        else:
            avg_time = times[0] if times else 0
            std_dev = 0

        # Escribir los resultados en el archivo
        outputFile.write(f"{size}\t{avg_time}\t{std_dev}\n")

        print(f"Experimento con tamaño de entrada {size} completado. Promedio de tiempo: {avg_time:.6f} segundos, Desviación estándar: {std_dev:.6f}")

print("Todos los experimentos completados y resultados guardados en 'bubble_sort_times.txt'")
